src/sage/libs/singular/ring.pyx

```diff
@@ -377,39 +497,56 @@ cdef ring *singular_ring_new(base_ring, n, names, term_order) except NULL:
         _cf = nInitChar( n_Z, NULL) # integer coefficient ring
         _ring = rDefault (_cf, nvars, _names, nblcks, _order, _block0, _block1, _wvhdl)
 
-    elif (isinstance(base_ring, FiniteField_generic) and base_ring.is_prime_field()):
-        if base_ring.characteristic() <= 2147483647:
+    elif isinstance(base_ring, sage.rings.abc.IntegerModRing):
+
+        ch = base_ring.characteristic()
+        if ch < 2:
+            raise NotImplementedError(f"polynomials over {base_ring} are not supported in Singular")
+
+        isprime = ch.is_prime()
+
+        if isprime and ch <= 2147483647 and isinstance(base_ring, FiniteField_generic):
+            # don't use this branch for e.g. Zmod(5)
             characteristic = base_ring.characteristic()
+
+            # example for simpler ring creation interface without monomial orderings:
+            #_ring = rDefault(characteristic, nvars, _names)
+
+            _ring = rDefault(characteristic, nvars, _names, nblcks, _order, _block0, _block1, _wvhdl)
+
         else:
-            raise TypeError("Characteristic p must be <= 2147483647.")
+            modbase, cexponent = ch.perfect_power()
 
-        # example for simpler ring creation interface without monomial orderings:
-        #_ring = rDefault(characteristic, nvars, _names)
+            if modbase == 2 and cexponent > 1:
+                _cf = nInitChar(n_Z2m, <void *>cexponent)
 
-        _ring = rDefault(characteristic, nvars, _names, nblcks, _order, _block0, _block1, _wvhdl)
+            elif modbase.is_prime() and cexponent > 1:
+                _info.base = <__mpz_struct*>omAlloc(sizeof(__mpz_struct))
```

Comment:
```suggestion
                _info.base = <__mpz_struct *>omAlloc(sizeof(__mpz_struct))
```

Reply:
I just copy pasted the code around, but sure.